File: src/0121-Best-Time-to-Buy-and-Sell-Stock/0121.py
class Solution:
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        result, buy = 0, float('inf')
        for price in prices:
            if buy > price:
                buy = price
            
            if result < price - buy:
                result = price - buy

        return result

        
if __name__ == "__main__":
    prices = [7,1,5,3,6,4]
    print(Solution().maxProfit(prices))

File: src/0121-Best-Time-to-Buy-and-Sell-Stock/0121.cpp
#include <iostream>
#include <vector>
#include <cstdint>
using namespace std;

static int x = []() {std::ios::sync_with_stdio(false); cin.tie(0); return 0; }();
class Solution 
{
public:
    int maxProfit(vector<int>& prices) 
    {
        int result = 0;
        int buy = INT_MAX;
        for (auto price : prices)
        {
            if (buy > price) buy = price;
            if (result < price - buy) result = price - buy;
        }
        return result;

    }
};
int main()
{
    vector<int> prices = {7,1,5,3,6,4};
    cout << Solution().maxProfit(prices);
}